File: Two-pointers/31-nextPermutation/31.py
from typing import List

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        i = n - 1
        while i >= 1:
            if nums[i] > nums[i - 1]:
                for j in range(n - 1, i - 1, -1):
                    if nums[j] > nums[i - 1]:
                        nums[j], nums[i - 1] = nums[i - 1], nums[j]
                        break
                break
            i -= 1
        nums[i:] = reversed(nums[i:])


if __name__ == "__main__":
    l = list([2, 1, 3, 5])
    l[1:] = reversed(l[1:])
    print(l)

File: Two-pointers/31-nextPermutation/31.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

class Solution {
 public:
  void nextPermutation(vector<int>& nums) {
    // 1 2 3
    // 1 3 2
    // 2 1 3

    // 1 2 5 4 3 0
    // (1 3 5 4 2 0)
    // 1 3 0 2 4 5
    int n = nums.size();
    int i = n - 1;
    for (; i >= 1; --i) {
      if (nums[i] > nums[i - 1]) {
        // 找到第一个“较小数”：nums[i-1]
        for (int j = n - 1; j > i - 1; --j) {
          // 找到第一个小于 nums[i-1] 的“较大数”：nums[j]
          if (nums[j] > nums[i - 1]) {
            swap(nums[i - 1], nums[j]);  // 交换“较小数”和“较大数”
            break;
          }
        }
        break;
      }
    }
    // 此时 nums[i:n] 必为降序，需将其反转为升序。
    reverse(nums.begin() + i, nums.end());
  }
};

int main() {
  vector<int> test = {4, 3, 1, 2};
  Solution obj;
  obj.nextPermutation(test);
  for (int n : test) {
    cout << n << " ";
  }
  return 0;
}